### DISCUSSION TITLE
Unable to add logo space with PDF output

### DISCUSSION THREAD

**Discussion by reconstrukt** [chillerlan/php-qrcode#162] [General]
Hi,

I'm attempting to generate a QR Code with `addLogoSpace` option set to `true` , output to a PDF (`'outputType'   => QRCode::OUTPUT_FPDF,`.  But, receiving an error, details below.

Full code sample is below, w/ configuration options.  Pls note I am not placing a logo in this example, just attempting to add space for it in the matrix.  

I am using php v7.4.25 with the `"chillerlan/php-qrcode": "dev-main"` version in my composer.json. 

Thank you!
Matt

```
$dark = [255, 255, 255];      // white code 
$trans = [238, 35, 36];       // red background

$options = new QROptions([
    'version'      => 5,        
    'outputType'   => QRCode::OUTPUT_FPDF,
    'eccLevel'     => EccLevel::H,
    'addLogoSpace'    => true,
    'logoSpaceWidth'  => 13,
    'logoSpaceHeight' => 13,
    'scale'        => 5,
    'imageBase64'  => false,
    'maskPattern' => 4,
    'moduleValues' => [
        // finder
        QRMatrix::M_FINDER | QRMatrix::IS_DARK     => $dark, // dark (true)
        QRMatrix::M_FINDER_DOT | QRMatrix::IS_DARK => $dark, // finder dot, dark (true)
        QRMatrix::M_FINDER                         => $trans, // light (false), white is the transparency color by default            
        // alignment
        QRMatrix::M_ALIGNMENT | QRMatrix::IS_DARK  => $dark,
        QRMatrix::M_ALIGNMENT                      => $trans,
        // timing
        QRMatrix::M_TIMING | QRMatrix::IS_DARK     => $dark,
        QRMatrix::M_TIMING                         => $trans,
        // format
        QRMatrix::M_FORMAT | QRMatrix::IS_DARK     => $dark,
        QRMatrix::M_FORMAT                         => $trans,
        // version
        QRMatrix::M_VERSION | QRMatrix::IS_DARK    => $dark,
        QRMatrix::M_VERSION                        => $trans,
        // data
        QRMatrix::M_DATA | QRMatrix::IS_DARK       => $dark,
        QRMatrix::M_DATA                           => $trans,
        // darkmodule
        QRMatrix::M_DARKMODULE | QRMatrix::IS_DARK => $dark,
        // separator
        QRMatrix::M_SEPARATOR                      => $trans,
        // quietzone
        QRMatrix::M_QUIETZONE                      => $trans,
    ],
    
]);

try{
  $qr = (new QRCode($options));
} catch(\Throwable $e){
  exit($e->getMessage());
}

$filename = uniqid() . '.pdf';
$filepath = config('filesystems.disks.temp.root') . '/' . $filename;    

$data = 'https://github.com/chillerlan/php-qrcode';

$qr->render($data, $filepath);

echo $filepath;
```

And i'm receiving the following error:

```
   chillerlan\QRCode\Data\QRCodeDataException

  code length overflow. (548 > 368 bit)

  at vendor/chillerlan/php-qrcode/src/Data/QRData.php:173
    169▕                }
    170▕
    171▕                // overflow, likely caused due to invalid version setting
    172▕                if($this->bitBuffer->getLength() > $MAX_BITS){
  ➜ 173▕                        throw new QRCodeDataException(
    174▕                                sprintf('code length overflow. (%d > %d bit)', $this->bitBuffer->getLength(), $MAX_BITS)
    175▕                        );
    176▕                }
    177▕

      +5 vendor frames
  6   app/Library/LabelRenderer.php:175
      chillerlan\QRCode\QRCode::render()
```

**Comment by codemasher** [comment]
The error means that the string you try to encode does not fit into the QR Code. Try increasing the version number or set it to `Version::AUTO`.

**Reply by reconstrukt** [reply]
Roger! I'll try that now

**Comment by reconstrukt** [comment]
hmm @codemasher i'm not seeing a public `AUTO` const on `chillerlan\QRCode\Common\Version` .
...
tried blindly setting `version` to `10`, that worked!

is there any docs on what the versions mean, and how to determine the "best" version to used based on the length of the string being passed into the qr code? (fwiw, we're just embedding dynamic URLs, approx ~70 characters or so, nothing crazy). 

also, is there a way to set the color of the logo space?
see screenshot below, its a white square by default

thank you,
matt
![image](https://user-images.githubusercontent.com/1252601/200967237-0476f979-4365-4b28-99da-f13d419b995b.png)

**Reply by codemasher** [reply]
>  i'm not seeing a public `AUTO` const on `chillerlan\QRCode\Common\Version` .

Uhh, [it should be](https://github.com/chillerlan/php-qrcode/blob/54fc7009a332595afd5fa5f70c0ea747c42973eb/src/Common/Version.php#L20). But you can also just omit the version setting, as auto detection is [enabled by default](https://github.com/chillerlan/php-qrcode/blob/main/src/QROptionsTrait.php#L29).

edit:
> is there any docs on what the versions mean, and how to determine the "best" version to used based on the length of the string being passed into the qr code?

There's nothing else to do on your end - the auto setting does that all for you - even for mixed mode QR Codes.
Version (and general QR Code info is here: https://www.qrcode.com/en/about/version.html )

> also, is there a way to set the color of the logo space?

I've added a "background" option recently but it's inconsitent across the several output modules for various reasons. Since you're using FPDF, you may consider their documentation (you can return the FPDF instance by setting `QROptions::returnResource` to true).

**Reply by codemasher** [reply]
Ok, i've added support for the `bgColor` and `drawLightModules` options in https://github.com/chillerlan/php-qrcode/commit/00f47b40f07d8c23f79041a435658f9da82a62ee so you can omit the module values for the light pixels.

**Comment by reconstrukt** [comment]
Thanks @codemasher for all your help! Will look into manually setting the background color on this end

**Reply by codemasher** [reply]
Um hey, wait a second - have you tried to set the quiet zone to dark? I just noticed in your code snippet above it is set to transparent. This should do the trick as the output module doesn't care for the given values.

Edit: oh nevermind - this *is* the orange color :)

**Reply by codemasher** [reply]
Excuse me, my brain is not working at full capacity right now... so the setting for the logo space would be `QRMatrix::M_LOGO`, which you would need to add to the module settings.

Also, I'll look into the FPDF output module and see if i can get the background setting working.

**Comment by reconstrukt** [comment]
@codemasher lol, no problem
just so i'm tracking, what should i try doing with ` QRMatrix::M_LOGO,` ?

**Reply by codemasher** [reply]
In your OP you see the array where you define the colors for the several matrix parts, just add a line there:
```php
        QRMatrix::M_LOGO => $trans,
```

**Comment by reconstrukt** [comment]
ah got it, copy that; will try now

**Comment by reconstrukt** [comment]
boom! worked like a champ
![image](https://user-images.githubusercontent.com/1252601/201136055-bab84d21-9d20-442c-b2e8-e84748170972.png)

**Reply by codemasher** [reply]
yay 🎉

**Reply by codemasher** [reply]
Hmm, thinking about a way to simply specify a single fill color for dark and light that works across all output modules in the same way to avoid the advanced `moduleValues` setting - currently there's only `markupDark` and `markupLight`.

**Reply by reconstrukt** [reply]
yeah that'd be cool. we generate a ton of these tags, all in different colors: some are light-on-dark, others dark-on-light, so we use those variables `$trans` and `$dark` with that single output template to generate each accordingly

**Reply by reconstrukt** [reply]
also @codemasher when you say "look into the FPDF output module and see if i can get the background setting working" does that mean there'd be a single prop we can pass a background image (resource or url?) to? should i standby for that?

**Reply by codemasher** [reply]
I assume you need the PDF format for printing purposes? I'd suggest using the SVG output instead which is way more flexible and allows for really fancy things and you can still add it to a PDF in case you really need to.

**Reply by codemasher** [reply]
> also @codemasher when you say "look into the FPDF output module and see if i can get the background setting working" does that mean there'd be a single prop we can pass a background image (resource or url?) to? should i standby for that?

By "background" i mean an opaque background color that fills the base layer before the individual QR pixels are drawn, so that any transparent areas will automatically colored.
SVG would also allow for background images, see #150 for example.

**Reply by reconstrukt** [reply]
@codemasher bingo, we're using PDF for printing; these QR codes are printed in a facility (at 2 sizes) and finished onto a roll of stickers/tags. 

re: background fill, got it, following.

so you're suggesting: 
 (a) render the QR out to SVG, which allows for more options (like more easily adding a logo image to center); then,
 (b) add that SVG into our final PDF template for printing (as usual)
 
 ?

**Reply by codemasher** [reply]
> so you're suggesting: 
 (a) render the QR out to SVG, which allows for more options (like more easily adding a logo image to center); then,
 (b) add that SVG into our final PDF template for printing (as usual)

Yes, this exactly. Converting SVG to PDF (or alternatively raster formats like PNG and BMP) can be done via the [Inkscape command line tool under linux](https://superuser.com/questions/381125/how-do-i-convert-an-svg-to-a-pdf-on-linux) (and probably also on Windows). There may also be other tools that allow converting or importing SVG into PDF, [TCPDF for example](https://tcpdf.org/examples/example_058/)

**Reply by reconstrukt** [reply]
we're using FPDF, but neither FPDF nor TCPDF have the ability to import an SVG. 
we may need to maintain the logo in PDF and import into the template as an additional step in the pipeline

**Reply by codemasher** [reply]
The TCPDF example i linked above shows SVG import. Either way, it's all just a suggestion - do whatever works best for you. :)

**Reply by reconstrukt** [reply]
copy, taking a look at that snip (circa 2013 lol)

**Reply by codemasher** [reply]
I agreee, TCPDF is a moloch and i wouldn't poke it with a stick... :D

**Reply by reconstrukt** [reply]
ok, so we ended up managing the logo assets as PDFs; and adding the logo to the center of the QR code (as another layer on the PDF) at render-time. just another step in tag creation. works great, little annoying to manage all the separate logo assets (diff colors), but lot better than bringing in TCPDF. :) 

thanks for your help with all the options above! cheers

**Reply by codemasher** [reply]
Sounds great! Cheers!
